src/trace_ng/log.zig

```diff
@@ -0,0 +1,1048 @@
+pub fn ScoppedLogger(comptime scope: ?[]const u8) type {
+    return union(LogKind) {
+        const Self = @This();
+        standard: *StandardErrLogger(scope),
+        testing: *TestingLogger(scope),
+        noop: void,
+        pub fn init(config: Config) !Self {
+            switch (config.kind) {
+                .standard => {
+                    return .{ .standard = try StandardErrLogger(scope).init(.{
+                        .allocator = config.allocator,
+                        .max_level = config.max_level,
+                        .max_buffer = config.max_buffer,
+                    }) };
+                },
+                .testing, .noop => {
+                    return .{ .testing = TestingLogger(scope).init(.{
+                        .allocator = config.allocator,
+                        .max_level = config.max_level,
+                        .max_buffer = config.max_buffer,
+                    }) };
+                },
+            }
+        }
+
+        pub fn deinit(self: *const Self) void {
+            switch (self.*) {
+                .standard => |*logger| {
+                    logger.*.deinit();
+                },
+                .testing => |*logger| {
+                    logger.*.deinit();
+                },
+                .noop => {},
+            }
+        }
+
+        pub fn unscoped(self: *const Self) !Logger {
+            switch (self.*) {
+                .standard => |logger| {
+                    return Logger.init(.{
+                        .allocator = logger.*.allocator,
+                        .max_buffer = logger.*.max_buffer,
+                        .kind = LogKind.standard,
+                    });
+                },
+                .testing => |logger| {
+                    return Logger.init(.{
+                        .allocator = logger.*.allocator,
+                        .kind = LogKind.testing,
+                    });
+                },
+                .noop => {
+                    @panic("Cannot scope noop");
+                },
+            }
+        }
+
+        pub fn withScope(self: *const Self, comptime new_scope: []const u8) !ScoppedLogger(new_scope) {
+            switch (self.*) {
+                .standard => |*logger| {
+                    return ScoppedLogger(new_scope).init(.{
+                        .allocator = logger.*.allocator,
+                        .max_buffer = logger.*.max_buffer,
+                        .kind = LogKind.standard,
+                    }) catch @panic("message: []const u8");
+                },
+                .testing => |*logger| {
+                    return ScoppedLogger(new_scope).init(.{
+                        .allocator = logger.*.allocator,
+                        .kind = LogKind.testing,
+                    }) catch @panic("message: []const u8");
+                },
+                .noop => {
+                    @panic("Cannot scope noop");
+                },
+            }
+        }
+
```

Comment:
Why not just use the existing logger instead of creating a new one? Also what's wrong with scoping noop?

Reply:
> Why not just use the existing logger instead of creating a new one? 

That is due to the pattern of passing in a logger, and allowing the struct use that to create a new one that is scoped.

```zig
    const StuffChild = struct {
        const StuffChild = @This();
        logger: ScopedLogger(@typeName(StuffChild)), // <- Scoped logger needed by struct

        pub fn init(logger: *const Logger) StuffChild {
            return .{ .logger = logger.withScope(@typeName(StuffChild)) catch { // <- New scoped logger created
                @panic("Init logger failed");
            } };
        }
```

Also if this is not done this way, ie, if the struct uses existing one, won't there be the risk of changing the scope externally?

> Also what's wrong with scoping noop?

Nothing. It could technically be scoped, just did not pay much attention to it yet and defaulted to panic. But can update.

Reply:
The instance of the underlying logger implementation can remain the same while changing the scope.

```zig
        pub fn unscoped(self: Self) Logger {
            return switch (self) {
                .standard => |logger| .{ .standard = logger },
                .testing => |logger| .{ .testing = logger },
                .noop => .noop,
            };
        }

        pub fn withScope(self: Self, comptime new_scope: []const u8) !ScoppedLogger(new_scope) {
            return switch (self) {
                .standard => |logger| .{ .standard = logger },
                .testing => |logger| .{ .testing = logger },
                .noop => .noop,
            };
        }
```